File: user/lib/time.c
/* SPDX-License-Identifier: GPL-2.0-only */
/*
 * Keira Kernel - Operating System Kernel
 * Copyright (C) 2026 Moh. Ananda Firmansyah Putra
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 */

/**
 * Keira Userland C Library: time.c
 *
 * C time management implementation using system calls sys_uptime and sys_sleep.
 */

#include "../include/time.h"

time_t time(time_t *tloc) {
    time_t now = (time_t)(sys_uptime() / 1000);
    if (tloc) {
        *tloc = now;
    }
    return now;
}

clock_t clock(void) {
    return (clock_t)sys_uptime();
}

double difftime(time_t time1, time_t time0) {
    return (double)(time1 - time0);
}

unsigned int sleep(unsigned int seconds) {
    sys_sleep((unsigned long)seconds * 1000);
    return 0;
}

File: crates/shell/src/cmds/time.rs
// SPDX-License-Identifier: GPL-2.0-only
//
// Keira Kernel - Operating System Kernel
// Copyright (C) 2026 Moh. Ananda Firmansyah Putra
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.

#![allow(unused_variables, unused_unsafe)]

//!
//! Implementation of the 'time' shell command.

use crate::executor::*;
use keira_io::vga;

pub fn run(parts: &mut core::str::SplitWhitespace) {
    unsafe {
        if let Some("-h") | Some("--help") = parts.next() {
            vga::print_str("Usage: time\n\n");
            vga::print_str("Description:\n  Query CMOS Real-Time Clock (RTC) hardware registers to display system date and UTC time.\n\n");
            vga::print_str("Options:\n  -h, --help    Show this help message and exit\n");
            return;
        }

        let mut time = RtcTime {
            second: 0,
            minute: 0,
            hour: 0,
            day: 0,
            month: 0,
            year: 0,
        };
        unsafe {
            rtc_get_time(&mut time as *mut RtcTime);
        }
        vga::set_color(vga::Color::White, vga::Color::Black);
        vga::print_str("Date: ");
        vga::set_color(vga::Color::White, vga::Color::Black);
        vga::print_u64(time.year as u64);
        vga::print_str("-");
        print_2digit(time.month as u64);
        vga::print_str("-");
        print_2digit(time.day as u64);
        vga::print_str(" ");
        print_2digit(time.hour as u64);
        vga::print_str(":");
        print_2digit(time.minute as u64);
        vga::print_str(":");
        print_2digit(time.second as u64);
        vga::print_str(" UTC\n");
        vga::set_color(vga::Color::LightGrey, vga::Color::Black);
    }
}
